### DISCUSSION TITLE
How to improve response speed of the authentication request with introspection-uri

### DISCUSSION THREAD

**Discussion by baixinsui** [zitadel/zitadel#6310] [Q&A] [answered]
The response speed of calling introspection-uri [http://localhost:8080}/oauth/v2/introspect) is slow, The authentication request may take 1200 ~ 2500 ms,  but the health request only takes 2 ms. How to improve response speed of the authentication request?
![image](https://github.com/zitadel/zitadel/assets/121531362/64e4721c-d1a4-4d82-b63f-6e08fe3b22d3)

**Comment by fforootd** [comment, marked as answer]
You have multiple options:

1) you can reduce the cost of the bcrypt implementation https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L367 to something like 11 or 12
2) you could use private/public key authentication against the introspection endpoint instead of a password (client secret) https://zitadel.com/docs/guides/integrate/token-introspection/private-key-jwt#register-the-api-in-zitadel-and-generate-private-and-public-keys

Hope this helps 😁

**Reply by forsak3n** [reply]
I'm using argon2id for password hashing and it doesn't have `Cost` configuration and I can't use public key authentication. What are my options?

Under load (130rps) introspection could take more than 10s to complete, even on a beefy machine

**Reply by fforootd** [reply]
@forsak3n please share your config and hardware in more detail?

Such a high latency indicates another issue IMO.

**Reply by forsak3n** [reply]
we have three instances, with zitadel and cockroach colocated, 4c/20ram.

and this is how it looks like when I'm enabling introspection, the graph is more or less the same on all three instances

![image](https://github.com/zitadel/zitadel/assets/2914046/40eb7da9-be52-4823-be6a-aab57a0a7553)
![image](https://github.com/zitadel/zitadel/assets/2914046/963b2d5f-597b-4d0b-8cff-0e9ba8cbab62)

Many requests are getting canceled by NGINX, because of the 10s timeout, those that are getting through are still in 9-10s range

EDIT: the config is mostly default, with the exception of password hashing

```yaml
    Hasher:
      Algorithm: "argon2id"
      Time: 1
      Memory: 65536
      Threads: 4
```
Let me know if you need something more specific

**Reply by forsak3n** [reply]
Config in its entirety

https://pastebin.com/DM2BPUg2

**Reply by fforootd** [reply]
Thanks for the data, what version of zitadel are you running?

**Reply by forsak3n** [reply]
I was running 2.43.0, but updated to 2.43.11 to see whether that would make a difference

**Reply by fforootd** [reply]
Got it. 

Looking at the hasher config it still looks like an expensive config. Have you tried playing around with the values?
ATM you will need to recreate the client secret to update it to the new config (after you changed the config).

In the coming release (2.50) I believe we introduce a separated config for machine users https://github.com/zitadel/zitadel/blob/main/cmd/defaults.yaml#L484

**Reply by forsak3n** [reply]
I will give it a try, thanks! 

And just to be clear, this level of performance is "expected", because of the password hashing?

**Reply by fforootd** [reply]
In short no 😁 it should be fast

I think we see two problems here:

1) The secret hashing takes a lot of time (try tinkering with the values)
2) The long time for hashing blocks request to the database and plagues it to some extend

Maybe also try upgrading to the newest version (2.49) which contains a lot of perf fixes. The coming version 2.50 should bring a lot of perf fixes to userinfo and maybe also introspect (maybe that happens in 2.51)

**Reply by forsak3n** [reply]
I see, thanks for the clarification.

The upgrade to 2.43.0 from 2.37.3 was rather painful, because of the event store changes (which introduced performance bottlenecks somewhere in 2.41.x, I think?), so I have to tread carefully now and be very conservative with the upgrades, unfortunately.

**Reply by fforootd** [reply]
I feel you 2.39 was a problem. But we are aware of what we did there. Usually it should be more relaxed from no on forward or at least we have not seen big problems.

**Reply by forsak3n** [reply]
Not sure if this was enough, but alas, no discernible difference.

```yaml
    Hasher:
      Algorithm: "argon2id"
      Time: 1
      Memory: 16384
      Threads: 1
```
![image](https://github.com/zitadel/zitadel/assets/2914046/665676af-8474-405d-ad84-2da84e1af100)

Could anything be gleaned from the traces if I enable it?

**Reply by fforootd** [reply]
Hm yeah, traces might help to figure where we see the bottleneck.

Just that you are aware we released 2.50.0 today, which can automatically update the secrets for you and you can configure a separate hasher setting for human and machines.

I think looking at the hasher settings that the DB must have some issues with this.

**Reply by forsak3n** [reply]
So, I bit the bullet and upgraded to 2.50.0

While it somewhat helped with the latency itself, I think it also introduced (or boosted) another problem. 

Now I'm getting this message multiple times per second, once the load increases. I'm not sure how, but it also breaks session flows (create/update/terminate) which precludes users from logging in (we are using custom UI).

```
{"caller":"/home/runner/work/zitadel/zitadel/internal/command/command.go:299","error":"retrying txn failed after 50 attempts. original error: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: WriteTooOldError: write for key /Table/7421/1/\"234593317696831512\"/\"project\"/\"234605784107843608\"/1320927/0 at timestamp 1713275705.687786871,0 too old; wrote at 1713275705.735846522,1: \"sql txn\" meta={id=03d1e254 key=/Table/7421/1/\"234593317696831512\"/\"project\"/\"234605784107843608\"/1320926/0 pri=0.01851213 epo=0 ts=1713275705.735846522,1 min=1713275705.687786871,0 seq=5} lock=true stat=PENDING rts=1713275705.687786871,0 wto=false gul=1713275706.187786871,0 (SQLSTATE 40001).","level":"error","msg":"could not push event \"project.application.api.secret.check.succeeded\"","time":"2024-04-16T13:55:05Z"}
```

And in app I'm getting a bunch of "Errors.PermissionDenied (QUERY-dsfr3)" intermittently while using a perfectly valid session token.

All of this because of the measly 70rps.

Once I disable introspection checks, problem goes away, users can log in again. For what it's worth, I think introspection with basic auth is not suitable for production workloads :(

**Reply by fforootd** [reply]
Hm that is really strange. 

We have a huge improvement coming for introspection in the next release (fingers crossed).

To me it sounds like you have see an "ahead of line blocking" issue.

Would you be up for a call to show us around your setup? Maybe we can spot the problem.

**Reply by adlerhurst** [reply]
Hi we released v2.50.2 which should at least fix the retry error
